### PR TITLE
executor,kv: support timebounded staleness transaction

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/19817

Problem Summary:
Support time bounded staleness transaction.

### What is changed and how it works?

What's Changed:
Support time bounded  staleness transaction by:

1. maintaining resolveTS for each store in an extra goroutine.
2. support time-bounded staleness transaction in `executeBegin`

### Check List 

Tests 

- Unit test

I have tested it with https://github.com/tikv/tikv/pull/9791, it seems work well.

### Release note 

* No release note

### CODE REVIEW

File: executor/executor_test.go
Comment (Yisaer) [inline executor/executor_test.go:7814]: I moved all the staleness txn unit test to a dedicated test file from executor_test

File: store/tikv/kv.go
Comment (nolouch) [inline store/tikv/kv.go:139]: Can we reduce the concept of `xxxxts`? There are already have `start ts`, `commit ts`, `commit ts`, `read ts`, `resolve ts` and so on, maybe directly name it as `resolve ts`.

File: store/tikv/kv.go
Comment (xhebox) [inline store/tikv/kv.go]: I think `getGlobalMinSafeTS` and `getMinSafeTSByStores` can be merged into one? You could just pass `s.regionCache.getStoresByType(tikvrpc.TiKV)` to `getMinSafeTSByStores`.

File: store/tikv/kv.go
Comment (Yisaer) [reply store/tikv/kv.go]: updated.

**Review by xhebox — COMMENTED** [review]
Rest LGTM

File: executor/stale_txn_test.go
Comment (JmPotato) [inline executor/stale_txn_test.go:104]: Why `±2` here?

**Review by JmPotato — CHANGES_REQUESTED** [review]

File: executor/stale_txn_test.go
Comment (Yisaer) [reply executor/stale_txn_test.go:104]: stale txn tolerate 2 seconds deviation for startts.

File: store/tikv/kv.go
Comment (Yisaer) [reply store/tikv/kv.go]: `getGlobalMinSafeTS` removed

File: executor/stale_txn_test.go
Comment (Yisaer) [reply executor/stale_txn_test.go:94]: we will address this in next pr.

File: executor/stale_txn_test.go
Comment (JmPotato) [reply executor/stale_txn_test.go:104]: What about setting it as a const variable?

Comment (Yisaer) [reply executor/stale_txn_test.go:105]: updated.

Comment (xhebox) [reply executor/stale_txn_test.go:105]: @JmPotato gentle ping

**Review by JmPotato — APPROVED** [review]

File: executor/stale_txn_test.go
Comment (JmPotato) [reply executor/stale_txn_test.go:105]: Approved, but it seems that I don't have LGTM access in `sig-execution` or `sig-sql-infa`.

File: store/tikv/kv.go
Comment (djshow832) [inline store/tikv/kv.go:231]:
```suggestion
	// If the resolveTS is larger than the minStartTS, we will use resolveTS as StartTS, otherwise we will use
```

File: store/tikv/kv.go
Comment (djshow832) [inline store/tikv/kv.go:242]: Why not compare with resolvedTS?

File: store/tikv/kv.go
Comment (djshow832) [inline store/tikv/kv.go:229]: What if `stores` is empty? The resolveTS will be MaxUint64.

File: store/tikv/kv.go
Comment (djshow832) [inline store/tikv/kv.go:432]: All stores? Why not the stores in the same zone?

File: store/tikv/kv.go
Comment (djshow832) [inline store/tikv/kv.go:449]: What will happen if TiKV is an older version, which cannot provide safeTS?

File: store/tikv/kv.go
Comment (djshow832) [inline store/tikv/kv.go:413]: What if:
1. A new store is just added into the zone.
2. The store just begins to replicate data to itself.
3. The minSafeTS is very stale, and TiDB has to read with minStartTS.
4. When TiDB requests data from the new store with minStartTS, the store will always fail until it catches up with the leader.

File: store/tikv/oracle/oracle.go
Comment (djshow832) [inline store/tikv/oracle/oracle.go:155]: What's the difference with just comparing the int value of `tsoOne` and `tsoTwo`?

File: store/tikv/kv.go
Comment (djshow832) [reply store/tikv/kv.go:413]: Oh, so the replica is a learner, and TiDB won't read it.

File: store/tikv/kv.go
Comment (Yisaer) [reply store/tikv/kv.go:243]: `BeginWithMaxPrevSec` will directly call `BeginWithMinStartTS` after getting minStartTS, and the reslovedTS will be compared with minStartTS in `BeginWithMinStartTS`

File: store/tikv/kv.go
Comment (Yisaer) [reply store/tikv/kv.go:230]: updated, If the stores is empty, we will directly return 0 in order to let txn use minStartTS as startTS.

File: store/tikv/kv.go
Comment (Yisaer) [reply store/tikv/kv.go:437]: When `txnScope` is global, we need to startTS with all the stores resolvedTS, so we need to maintian all the resolvedTS for each tikv.

File: store/tikv/oracle/oracle.go
Comment (Yisaer) [reply store/tikv/oracle/oracle.go:155]: The logic is migrated from PD, I think there is no difference.

File: store/tikv/kv.go
Comment (Yisaer) [reply store/tikv/kv.go:418]: If the resolveTS is too stale, we will use minStartTS directly.

File: store/tikv/kv.go
Comment (Yisaer) [reply store/tikv/kv.go:454]: The deployment tool will upgrade tikv first before tidb, and PD will prevent the lower version tikv to join the cluster. Currently if the tikv couldn't provide safeTS for itself, the reslovedTS maintained in tidb will be zero, and the startTS of the txn will be fallbacked to the minStartTS.

File: store/tikv/kv.go
Comment (Yisaer) [reply store/tikv/kv.go:216]: Not possible, the value is returned from `CheckAndGetTxnScope`.

**Comment by Yisaer** [comment]
We need to wait https://github.com/tikv/tikv/pull/9385 to be merged, then merge this pr.

**Comment by hi-rustin** [comment]
@Yisaer  Silence the bot with the /merge cancel comment for consistent failures.

**Comment by hi-rustin** [comment]
A PR for a draft that can be merged looks very strange, so I suggest you unmerge first until another PR lands.

**Comment by hi-rustin** [comment]
> We need to wait [tikv/tikv#9385](https://github.com/tikv/tikv/pull/9385) to be merged, then merge this pr.

You can use the /hold command to prevent the merging of PR. No need to convert to draft.

**Pull request merged**
